object: adding ceph cosi driver (#12415)
rook/rook: Documentation/Storage-Configuration/Object-Storage-RGW/object-storage.md

<?xml version="1.0" encoding="UTF-8"?><ListAllMyBucketsResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Owner><ID>anonymous</ID><DisplayName></DisplayName></Owner><Buckets></Buckets></ListAllMyBucketsResult>
```

> ## Create a Bucket via lib-bucket-provisioner and Object Bucket Claim

Reviewer: How about we keep the title the same, and then have an info label like the following? Everything about COSI (except this note) would just be covered in the new document. When COSI is no longer experimental, then we would update this document to use COSI by default. 
```suggestion
## Create a Bucket

!!! info
    This document will guide you to create a bucket with an Object Bucket Claim (OBC). To create a bucket with the experimental COSI Driver, see the [COSI documentation](cosi.md). 
```

## Create a Bucket

!!! info

>     This document will guide you to create a bucket with an Object Bucket Claim (OBC). To create a bucket with the experimental COSI Driver, see the [COSI documentation](cosi.md).

Reviewer 2: Reword to avoid "you"

```suggestion
    This document is a guide for creating bucket with an Object Bucket Claim (OBC). To create a bucket with the experimental COSI Driver, see the [COSI documentation](cosi.md).
```

cat /tmp/rookObj-download
```

### Create a Bucket via Ceph COSI Driver

The Ceph COSI driver will be started automatically with default settings when first Object Store gets created. The driver will be deleted when Rook Operator is uninstalled. The driver will be deployed in the same namespace as Rook operator. The COSI controller needs already running if the Ceph COSI Driver needs to come up. The COSI controller can be deployed by following the commands below:

```bash

> kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-api ## installs CRDs

Reviewer: Can we include them in the helm chart? Or is there a COSI helm chart that we could make a dependency? For experimental mode let's not worry about the helm chart, but when COSI is considered stable we need it to be considered via helm.

Author: I send out to PR to create a helm chart for the cosi controller, it's not review till now https://github.com/kubernetes-sigs/container-object-storage-interface-controller/pull/75. Even I like to do it via helm instead of using kubectl

kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-controller ## installs COSI controller
```

Now user can define BucketClass and BucketClaim to create a bucket. The BucketClass and BucketClaim can be defined as below:

```yaml
kind: BucketClass
apiVersion: objectstorage.k8s.io/v1alpha1
metadata:
  name: sample-bcc
driverName: ceph.objectstorage.k8s.io
deletionPolicy: Delete
parameters:
  ObjectStoreUserSecretName: rook-ceph-object-user-<store_name>-cosi
  ObjectStoreNamespace: rook-ceph
---
kind: BucketClaim
apiVersion: objectstorage.k8s.io/v1alpha1
metadata:
  name: sample-bc
spec:
  bucketClassName: sample-bcc
  protocol
```

> The `CephObjectStoreUser` named cosi will be automatically and user can refer to the secret name in the BucketClass. The BucketClass and BucketClaim can be created by using above the yaml file. After successful creation of above CRDs user can fetch them by using below commands:

Reviewer: Agreed that we don't expect Rook to create any CRs, since we expect the users to create them. Having Rook internally create the user and the secret makes sense. This can be implemented by the object store calling the object store user controller code for creating the built-in user and secret. This way we get the benefit of sharing code, but we don't need a CR to trigger it, or worry about the user deleting the CR.

> The `CephObjectStoreUser` named cosi will be automatically and user can refer to the secret name in the BucketClass. The BucketClass and BucketClaim can be created by using above the yaml file. After successful creation of above CRDs user can fetch them by using below commands:

Author: Will add the automatic creation as Update on the Initial PR

```bash

> $ kubectl get bucketclass,bucket,bucketclaim

Reviewer: I would suggest to remove lines 279-290. Rook docs don't usually show how to get the resources they just created unless there is some interesting status they need to see.
